### PR TITLE
Add reflect `ClassDef.apply` and `Symbol.newClass`

### PR DESCRIPTION
**Pull request by nicolasstucki**
We also add
* `SymbolMethods.typeRef`
* `SymbolMethods.termRef`
* `TypeTreeModule.ref`

### CODE REVIEW

File: library/src/scala/quoted/Quotes.scala [hunk 1]
```diff
@@ -464,6 +464,7 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
 
     /** Methods of the module object `val ClassDef` */
     trait ClassDefModule { this: ClassDef.type =>
+      @experimental def apply(cls: Symbol, parents: List[Tree /* Term | TypeTree */], body: List[Statement]): ClassDef
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:467]: There's no example where we pass an actual term (e.g. `new Foo("a")` to pass a parameter to the parent), is that actually supported?

**Review by smarter — CHANGES_REQUESTED** [review]
Otherwise LGTM.

File: compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala [hunk 2]
```diff
@@ -227,6 +227,11 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
     end ClassDefTypeTest
 
     object ClassDef extends ClassDefModule:
+      def apply(cls: Symbol, parents: List[Tree], body: List[Statement]): ClassDef =
+        val untpdCtr = untpd.DefDef(nme.CONSTRUCTOR, Nil, tpd.TypeTree(dotc.core.Symbols.defn.UnitClass.typeRef), untpd.EmptyTree)
```
[the remark this answers is not in the document]
Comment (nicolasstucki) [reply compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala]: Fixed

File: library/src/scala/quoted/Quotes.scala (hunk 1 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:467]: Added `tests/run-macros/newClassExtendsClassParams` to test this feature.

It actually interacts butefully with quoted expressions. We can define the parents as
```scala
  val parents = List('{ new Foo(1) }.asTerm)
```

File: tests/neg-macros/newClassExtendsNoParents/Macro_1.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+private def makeClassExpr(nameExpr: Expr[String])(using Quotes): Expr[Any] = {
+  import quotes.reflect.*
+
+  val name = nameExpr.valueOrAbort
+  val parents = List.empty[Tree] // BUG: first parent is not a class
+  def decls(cls: Symbol): List[Symbol] = Nil
+
+  val cls = Symbol.newClass(Symbol.spliceOwner, name, parents = Nil, decls, selfInfo = None)
+  val clsDef = ClassDef(cls, parents, body = List())
+  val newCls = Typed(Apply(Select(New(TypeIdent(cls)), cls.primaryConstructor), Nil), TypeTree.of[Object])
```
Comment (smarter) [inline tests/neg-macros/newClassExtendsNoParents/Macro_1.scala:13]: I guess this is orthogonal but in tpd we have a New overload that takes constructor arguments  this could be useful here too

**Review by smarter — CHANGES_REQUESTED** [review]

File: library/src/scala/quoted/Quotes.scala [hunk 4]
```diff
@@ -3533,19 +3534,73 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+       *    class `name`() extends Object with Foo {
+       *      def foo(): Unit = println("Calling foo")
+       *    }
+       *    new `name`(): Foo
+       *  }
+       *  ```
+       *
+       *  @param parent The owner of the class
+       *  @param name The name of the class
+       *  @param parents The parent classes of the class
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3567]:
```suggestion
       *  @param parents The parent classes of the class. The first parent must not be a trait.
```

File: library/src/scala/quoted/Quotes.scala [hunk 5]
```diff
@@ -3807,6 +3868,14 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
 
         /** Case class or case object children of a sealed trait or cases of an `enum`. */
         def children: List[Symbol]
+
+        /** Type reference to the symbol */
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3872]: Maybe mention that this is as seen from the owner:
```suggestion
        /** Type reference to the symbol usable in the scope of its owner. */
```

File: library/src/scala/quoted/Quotes.scala [hunk 6]
```diff
@@ -3533,19 +3534,73 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+       *  @param self The self type of the class if it has one
+       *
+       *  This symbol starts without an accompanying definition.
+       *  It is the meta-programmer's responsibility to provide exactly one corresponding definition by passing
+       *  this symbol to the ClassDef constructor.
+       *
+       *  @note As a macro can only splice code into the point at which it is expanded, all generated symbols must be
+       *        direct or indirect children of the reflection context's owner.
+       */
+      @experimental def newClass(parent: Symbol, name: String, parents: List[TypeRepr], decls: Symbol => List[Symbol], selfInfo: Option[TypeRepr]): Symbol
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3578]: Maybe call it selfType instead of selfInfo since that's how we refer to this feature in the language.
```suggestion
      @experimental def newClass(parent: Symbol, name: String, parents: List[TypeRepr], decls: Symbol => List[Symbol], selfType: Option[TypeRepr]): Symbol
```

File: library/src/scala/quoted/Quotes.scala [hunk 7]
```diff
@@ -464,6 +464,13 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
 
     /** Methods of the module object `val ClassDef` */
     trait ClassDefModule { this: ClassDef.type =>
+      /** Create a class definition tree
+       *
+       *  @param cls The class symbol
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:469]:
```suggestion
       *  @param cls The class symbol created using `Symbol.newClass`
```

File: library/src/scala/quoted/Quotes.scala [hunk 8]
```diff
@@ -464,6 +464,13 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
+       *  @param parents The parents trees class. The trees must align with the parent types of `cls`.
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:470]: This description should be expanded (or some examples should be added) to explain that the parents can either be TypeTrees or Terms, and the terms are `new` calls so that arguments can be passed to them.

**Review by smarter — APPROVED** [review]
Otherwise LGTM.

File: library/src/scala/quoted/Quotes.scala [hunk 9]
```diff
@@ -3807,6 +3874,14 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
 
         /** Case class or case object children of a sealed trait or cases of an `enum`. */
         def children: List[Symbol]
+
+        /** Type reference to the symbol usable in the scope of its owner. */
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3878]: This (and termRef) should be expanded to mention something like:
```suggestion
        /** Type reference to the symbol usable in the scope of its owner.
         *  To get a reference to a symbol from a specific prefix `tp`, use `tp.memberType(symbol)` instead.
         */
```

Comment (smarter) [reply library/src/scala/quoted/Quotes.scala:3878]: Also `@pre symbol.isType returns true`

File: library/src/scala/quoted/Quotes.scala (hunk 7 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:469]: Not necessarily, it can be the symbol of a class for which we are transforming the tree.

Comment (smarter) [reply library/src/scala/quoted/Quotes.scala:469]: OK, then the comment could just mention that Symbol.newClass should be used to create a new class symbol. Also if we're transforming a tree I think most of the time we should use ClassDef.copy instead of ClassDef.apply.

File: library/src/scala/quoted/Quotes.scala (hunk 9 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala]: Done

File: library/src/scala/quoted/Quotes.scala (hunk 8 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:470]: Done

File: library/src/scala/quoted/Quotes.scala (hunk 7 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala]: Done

File: library/src/scala/quoted/Quotes.scala [hunk 10]
```diff
@@ -2524,6 +2533,11 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
         def typeSymbol: Symbol
         def termSymbol: Symbol
         def isSingleton: Boolean
+
+       /** This type seen as if it were the type of a member of prefix type `self` declared in class `member.owner`.
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:2537]: I think this description is incorrect, we don't return a version of "This type" but of the type of the member.
```suggestion
       /** The type of `member` as seen from prefix `self`
        *
        *  @pre `member` is an accessible member of `self`
```

File: tests/neg-macros/newClassExtendsNoParents.check [hunk 11]
```diff
@@ -0,0 +1,19 @@
+
+-- Error: tests/neg-macros/newClassExtendsNoParents/Test_2.scala:1:25 --------------------------------------------------
+1 |def test: Any = makeClass("foo") // error
+  |                ^^^^^^^^^^^^^^^^
+  |                Exception occurred while executing macro expansion.
+  |                java.lang.AssertionError: assertion failed: First parent must be a class
+  |                	at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:8)
+  |                	at scala.quoted.runtime.impl.QuotesImpl$reflect$Symbol$.newClass(QuotesImpl.scala:2464)
```
Comment (smarter) [inline tests/neg-macros/newClassExtendsNoParents.check:8]: This checkfile is too specific: it would need to be updated every time we add/remove a line in QuotesImpl.scala (hence why the tests are failing right now).

File: library/src/scala/quoted/Quotes.scala [hunk 12]
```diff
@@ -3807,6 +3883,19 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
...
         def children: List[Symbol]
+
+        /** Type reference to the symbol usable in the scope of its owner.
+         *
+         *  To get a reference to a symbol from a specific prefix `tp`, use `tp.memberType(symbol)` instead.
+         *
+         *  @pre symbol.isType returns true
+         */
+        @experimental
+        def typeRef: TypeRef
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3894]: It seems that we're missing an API to create a TypeRef/TermRef from a specific prefix: with `foo.typeRef` I can get a TypeRef `this.foo` and with `pre.memberType(foo)` I can get the info corresponding to `pre.foo`, but is there a way to generate the TypeRef `pre.foo` itself?

File: tests/run-macros/newClassSelf/Macro_1.scala [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+/** Generate
+ *  ```
+ *  class `name`() extends Foo with Bar
+ * ```
+ */
+def makeFooBar(using Quotes)(name: String, fooCls: quotes.reflect.Symbol): quotes.reflect.ClassDef = {
+  import quotes.reflect.*
+  val parents = List(Inferred(fooCls.typeRef), TypeTree.of[Bar])
```
Comment (smarter) [inline tests/run-macros/newClassSelf/Macro_1.scala:53]: Perhaps instead of the `Inferred(foo.typeRef)` pattern we could have an overload of TypeTree.apply (or TypeTree.of? Not sure why we use `of` there) that takes a Symbol. This would be similar to the compiler-internal `tpd.ref` (in fact alternatively we could expose `ref` directly?)

**Review by smarter — CHANGES_REQUESTED** [review]

File: library/src/scala/quoted/Quotes.scala (hunk 12 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:3904]: Doesn't `TypeRepr.select` provide that?

Comment (smarter) [reply library/src/scala/quoted/Quotes.scala:3904]: Oh indeed, I missed that one, maybe each of their documentation should `@see` the other one.

File: tests/run-macros/newClassSelf/Macro_1.scala (hunk 13 above)
Comment (nicolasstucki) [reply tests/run-macros/newClassSelf/Macro_1.scala]: Added `TypeTree.apply`

File: library/src/scala/quoted/Quotes.scala (hunk 12 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala:3997]: Done

File: library/src/scala/quoted/Quotes.scala [hunk 14]
```diff
@@ -3604,6 +3679,9 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
       *
       *  Note: Also see reflect.let
       *
+      *  @param parent The owner of the /var/lazy val
```
Comment (smarter) [inline library/src/scala/quoted/Quotes.scala:3682]: Is this supposed to be:
```suggestion
      *  @param parent The owner of the val/var/lazy val
```

**Review by smarter — APPROVED** [review]
Otherwise LGTM!

File: library/src/scala/quoted/Quotes.scala (hunk 14 above)
Comment (nicolasstucki) [reply library/src/scala/quoted/Quotes.scala]: Fixed

**Comment by evbo** [comment]
@nicolasstucki thank you for this!!! I am noticing a class cast exception if I try pattern matching `typeRef` you added in this PR:

>java.lang.ClassCastException: class dotty.tools.dotc.core.Names$SimpleName cannot be cast to class dotty.tools.dotc.core.Names$TypeName (dotty.tool
s.dotc.core.Names$SimpleName and dotty.tools.dotc.cor
e.Names$TypeName are in unnamed module of loader sbt.
internal.classpath.ClassLoaderCache$Key$CachedClassLo
ader @2c470fec)

e.g. if I try to check the type of a case class field like:

```scala
// scala 3.2.0-RC2
declaredField.companionClass.caseFields.map { ccField =>
  val tt = ccField.typeRef match {
    case TypeRef(typeRepr, args) => s"args are: ${args}"
  }
}
```

I wasn't sure if this was a separate bug or not, happy to open a new one. It does seem odd seeing that exception; feels like a class cast exception happening internally so thought it might be important.
